stackslib/src/net/neighbors/rpc.rs

```diff
@@ -0,0 +1,285 @@
+                        &network.local_peer, event_id
+                    );
+                    return Ok(None);
+                } else {
+                    // conversation died
+                    debug!("{:?}: HTTP event {} hung up", &network.local_peer, event_id);
+                    return Err(NetError::PeerNotConnected);
+                }
+            };
+
+            // drive socket I/O
+            if let Some(request) = request_opt.take() {
+                convo.send_request(request)?;
+            };
+            HttpPeer::saturate_http_socket(socket, convo)?;
+
+            // see if we got any data
+            let Some(http_response) = convo.try_get_response() else {
+                // still waiting
+                debug!(
+                    "{:?}: HTTP event {} is still waiting for a response",
+                    &network.local_peer, event_id
+                );
+                return Ok(None);
+            };
+
+            Ok(Some(http_response))
+        })
+    }
+}
+
+/// This struct represents everything we need to iterate through a set of ongoing requests, in
+/// order to pull out completed replies.
+pub struct NeighborRPCMessageIterator<'a> {
+    network: &'a mut PeerNetwork,
+    neighbor_rpc: &'a mut NeighborRPC,
+}
+
+/// This is an iterator over completed requests
+impl Iterator for NeighborRPCMessageIterator<'_> {
+    type Item = (NeighborAddress, StacksHttpResponse);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut inflight = HashMap::new();
+        let mut ret = None;
+        let mut dead = vec![];
+        for (naddr, (event_id, mut request_opt)) in self.neighbor_rpc.state.drain() {
+            if ret.is_some() {
+                // just save for retry
+                inflight.insert(naddr, (event_id, request_opt));
+                continue;
+            }
+
+            let response =
+                match NeighborRPC::poll_next_reply(self.network, event_id, &mut request_opt) {
+                    Ok(Some(response)) => response,
+                    Ok(None) => {
+                        // keep trying
+                        inflight.insert(naddr, (event_id, request_opt));
+                        continue;
+                    }
+                    Err(NetError::WaitingForDNS) => {
+                        // keep trying
+                        inflight.insert(naddr, (event_id, request_opt));
+                        continue;
+                    }
+                    Err(_e) => {
+                        // declare this neighbor as dead by default
+                        dead.push(naddr);
+                        continue;
+                    }
+                };
+
+            ret = Some((naddr, response));
+        }
+        for naddr in dead.into_iter() {
+            self.neighbor_rpc.add_dead(self.network, &naddr);
+        }
+        self.neighbor_rpc.state.extend(inflight);
+        ret
```

Comment:
> If any request completes, ret is set to Some(), which is great, but why doesn't this function just immediately return at that point? (Or at least break the loop?).

I'd _love_ to break out the loop. The problem is that both `.drain()` and `.into_iter()` destroy the remaining contents of `self.neighbor_rpc.state`, so I have to gather them up and load them back into `self.neighbor_rpc.state` once a single response is found.

I suppose I could instead buffer up all responses from `poll_next_reply`, store them internally, and then just yield them on subsequent calls, but I don't think that addresses your concern either.

> Second, if every conversation is still in flight, the iterator returns None, which indicates that the iterator is finished, but that's not true here: the iterator actually isn't finished and it should be called again.

You should think of this iterator as an iterator over completed responses, not all responses (where "completed" means "completed at the time you call `.next()`").

> perhaps the return type changed so that you can indicate if the stream is finished (i.e., all conversations have completed) and should not be invoked again or if the stream just is waiting on results.

Callers don't actually use this iterator directly.  Instead, they use `NeighborRPC::collect_replies()` to get any new replies, and they use `NeighborRPC::count_inflight()` to see if there are any ongoing conversations.  These both do what you're suggesting.  Would you like me to make the iterator private?

Reply:
ping @kantai -- is it okay to leave this iterator code as-is, in light of the fact that it's really just used as a building block for the API that the downloader actually uses?

Reply:
I think it makes more sense for it to not implement Iterator at all, and just make this a normal instance method, where your return value can indicate whether or not there's still pending results or all inflight requests have completed. Is the trait generic actually used or is this the only implementation ever invoked by those methods?

Reply:
The latter

Reply:
I removed the iterator.  I also did so for `stackslib/src/net/neighbor/comms.rs`, which used a similar pattern (and was the inspiration for this code).